profiling/src/lib.rs

```diff
@@ -259,6 +268,56 @@ extern "C" fn minit(r#type: c_int, module_number: c_int) -> ZendResult {
         zend::zend_execute_internal = Some(execute_internal);
     };
 
+    #[cfg(feature = "allocation_profiling")]
+    {
+        let use_zend_alloc = std::env::var("USE_ZEND_ALLOC");
+        if !use_zend_alloc.is_ok() ||
+           use_zend_alloc.is_ok() && use_zend_alloc.ok().unwrap() == "1"
+        {
+            // check for neighboring custom memory handlers
+            unsafe {
+                zend::zend_mm_get_custom_handlers(
+                    zend::zend_mm_get_heap(),
+                    &mut PREV_CUSTOM_MM_ALLOC,
+                    &mut PREV_CUSTOM_MM_FREE,
+                    &mut PREV_CUSTOM_MM_REALLOC,
+                );
+            }
+
+            unsafe {
+                zend::zend_mm_set_custom_handlers(
+                    zend::zend_mm_get_heap(),
+                    Some(alloc_profiling_malloc),
+                    Some(alloc_profiling_free),
+                    Some(alloc_profiling_realloc),
+                );
+            }
+
+            let mut custom_mm_malloc: Option<zend::VmMmCustomAllocFn> = None;
+            let mut custom_mm_free: Option<zend::VmMmCustomFreeFn> = None;
+            let mut custom_mm_realloc: Option<zend::VmMmCustomReallocFn> = None;
+            unsafe {
+                zend::zend_mm_get_custom_handlers(
+                    zend::zend_mm_get_heap(),
+                    &mut custom_mm_malloc,
+                    &mut custom_mm_free,
+                    &mut custom_mm_realloc,
+                );
+            }
+            if custom_mm_free.is_none() ||
+                custom_mm_malloc.is_none() ||
+                custom_mm_realloc.is_none() ||
+                custom_mm_free != Some(alloc_profiling_free) ||
+                custom_mm_malloc != Some(alloc_profiling_malloc) ||
+                custom_mm_realloc != Some(alloc_profiling_realloc)
+            {
+                info!("Memory Allocation profiling could not be enabled. Pleas feel free to fill an issue stating the PHP version and installed modules.");
+            } else {
```

Comment:
The entire feature of a custom memory handler is bound to `ZEND_MM_CUSTOM` being defined:
https://heap.space/xref/PHP-8.1/Zend/zend_alloc.c?r=c7558e2f#2895
If this constant is initialised with `0` then a call to `zend_mm_set_custom_handlers` wont do anything. The only (IMHO) reliable way to find out if we actually succeeded is to check the pointers returned by `zend_mm_get_custom_handlers`. An alternative could be to call `is_zend_mm()` which returns `false` in case a custom memory handler is installed.

Reply:
Calling `is_zend_mm` sounds good to me. Also we should then actually note that in the info!() call, because this most likely means that they have a custom compiled build for extra performance, and just telling them ZEND_MM_CUSTOM being disabled is the most helpful thing to do.